File: release/purr-suite-skill-package-2026-03-21/scripts/lib/seller_bi.ts
import {
  DEFAULT_DB_PATH,
  ensureDatabase,
  queryAll,
  queryOne,
} from "../db/sqlite.js";
import { formatMinorAmount } from "./money.js";
import {
  addDays,
  formatDateOnly,
  parseDateOnly,
} from "./time.js";
import { isRecord } from "./types.js";

const COMMAND_ALIASES: Record<string, string> = {
  "seller-bi.sales_today": "seller_bi.sales_today",
  "seller-bi.revenue_this_month": "seller_bi.revenue_this_month",
};

const SELLER_BI_COMMANDS = new Set([
  "seller_bi.sales_today",
  "seller_bi.revenue_this_month",
]);

const SALES_COUNTABLE_ORDER_STATUSES = ["paid", "refunded", "fulfilled"] as const;

interface IdentityRow {
  id: number;
}

function invalid(status: string, reply: string): Record<string, unknown> {
  return { status, reply };
}

function normalizeText(value: unknown, fieldName: string): string {
  const normalized = String(value ?? "").trim();
  if (!normalized) {
    throw new Error(`${fieldName} is required`);
  }
  return normalized;
}

function requireParamsMapping(params: unknown): Record<string, unknown> {
  if (params === null || params === undefined) {
    return {};
  }
  if (!isRecord(params)) {
    throw new Error("params must be an object");
  }
  return { ...params };
}

function canonicalCommandCode(value: unknown): string {
  const normalized = String(value ?? "").trim().toLowerCase();
  return COMMAND_ALIASES[normalized] ?? normalized;
}

function parseAnchorDate(params: Record<string, unknown>): Date {
  const raw = normalizeText(params.anchor_date, "params.anchor_date");
  try {
    return parseDateOnly(raw, "params.anchor_date");
  } catch (error) {
    throw new Error("params.anchor_date must be an ISO date like YYYY-MM-DD");
  }
}

function dayWindow(anchorDate: Date): [string, string] {
  const start = `${formatDateOnly(anchorDate)} 00:00:00`;
  const end = `${formatDateOnly(addDays(anchorDate, 1))} 00:00:00`;
  return [start, end];
}

function monthWindow(anchorDate: Date): [string, string] {
  const year = anchorDate.getUTCFullYear();
  const monthIndex = anchorDate.getUTCMonth();
  const monthStart = new Date(Date.UTC(year, monthIndex, 1));
  const nextMonthStart =
    monthIndex === 11
      ? new Date(Date.UTC(year + 1, 0, 1))
      : new Date(Date.UTC(year, monthIndex + 1, 1));
  return [
    `${formatDateOnly(monthStart)} 00:00:00`,
    `${formatDateOnly(nextMonthStart)} 00:00:00`,
  ];
}

function fetchActiveOwner(
  db: ReturnType<typeof ensureDatabase>,
  options: { channel: string; external_user_id: string },
): IdentityRow | null {
  return queryOne<IdentityRow>(
    db,
    `
      SELECT *
      FROM identities
      WHERE channel = ? AND external_user_id = ? AND is_active = 1 AND role = 'owner'
      LIMIT 1
    `,
    [options.channel, options.external_user_id],
  );
}

function salesToday(
  db: ReturnType<typeof ensureDatabase>,
  anchorDate: Date,
): Record<string, unknown> {
  const [windowStart, windowEndExclusive] = dayWindow(anchorDate);
  const row = queryOne<{ sales_count: number }>(
    db,
    `
      SELECT COUNT(*) AS sales_count
      FROM orders
      WHERE status IN (?, ?, ?)
        AND paid_at IS NOT NULL
        AND paid_at >= ?
        AND paid_at < ?
    `,
    [
      SALES_COUNTABLE_ORDER_STATUSES[0],
      SALES_COUNTABLE_ORDER_STATUSES[1],
      SALES_COUNTABLE_ORDER_STATUSES[2],
      windowStart,
      windowEndExclusive,
    ],
  );
  const salesCount = row?.sales_count ?? 0;
  return {
    status: "computed",
    reply: `Sales count for ${formatDateOnly(anchorDate)} is ${salesCount}.`,
    metric_code: "seller_bi.sales_today",
    anchor_date: formatDateOnly(anchorDate),
    window_start: windowStart,
    window_end_exclusive: windowEndExclusive,
    sales_count: salesCount,
  };
}

function revenueThisMonth(
  db: ReturnType<typeof ensureDatabase>,
  anchorDate: Date,
): Record<string, unknown> {
  const [windowStart, windowEndExclusive] = monthWindow(anchorDate);
  const month = formatDateOnly(anchorDate).slice(0, 7);
  const rows = queryAll<{ currency: string; amount_minor: number }>(
    db,
    `
      SELECT currency, SUM(amount_minor) AS amount_minor
      FROM payments
      WHERE status = 'paid'
        AND paid_at IS NOT NULL
        AND paid_at >= ?
        AND paid_at < ?
      GROUP BY currency
      ORDER BY currency ASC
    `,
    [windowStart, windowEndExclusive],
  );

  const revenueRows = rows.map((row) => ({
    currency: String(row.currency),
    amount_minor: row.amount_minor,
    display_amount: formatMinorAmount(row.amount_minor, row.currency),
  }));

  const currencyCount = revenueRows.length;
  let reply = `No net paid revenue found for ${month}.`;
  if (currencyCount === 1) {
    reply = `Revenue for ${month} is ${revenueRows[0]!.display_amount}.`;
  } else if (currencyCount > 1) {
    reply = `Revenue for ${month} spans ${currencyCount} currencies.`;
  }

  return {
    status: "computed",
    reply,
    metric_code: "seller_bi.revenue_this_month",
    anchor_date: formatDateOnly(anchorDate),
    month,
    window_start: windowStart,
    window_end_exclusive: windowEndExclusive,
    currency_count: currencyCount,
    multi_currency: currencyCount > 1,
    revenue_rows: revenueRows,
  };
}

export function handleSellerBi(
  context: Record<string, unknown>,
): Record<string, unknown> {
  const commandCode = canonicalCommandCode(context.command_code);
  if (!SELLER_BI_COMMANDS.has(commandCode)) {
    return invalid("invalid_intent", `Unsupported seller-bi intent: ${commandCode}`);
  }

  const user = isRecord(context.user) ? context.user : {};
  const params = context.params;
  const channel = String(context.channel ?? "telegram").trim() || "telegram";
  const externalUserId = String(user.external_user_id ?? "").trim();
  if (!externalUserId) {
    return invalid("invalid_input", "user.external_user_id is required");
  }

  let normalizedParams: Record<string, unknown>;
  let anchorDate: Date;
  try {
    normalizedParams = requireParamsMapping(params);
    anchorDate = parseAnchorDate(normalizedParams);
  } catch (error) {
    return invalid("invalid_input", error instanceof Error ? error.message : String(error));
  }

  const runtime = isRecord(context.runtime) ? context.runtime : {};
  const dbPath = String(runtime.db_path ?? DEFAULT_DB_PATH);
  const db = ensureDatabase(dbPath);

  try {
    const owner = fetchActiveOwner(db, {
      channel,
      external_user_id: externalUserId,
    });
    if (owner === null) {
      return invalid("forbidden", "Only the active owner can query seller-bi metrics.");
    }

    if (commandCode === "seller_bi.sales_today") {
      return salesToday(db, anchorDate);
    }
    if (commandCode === "seller_bi.revenue_this_month") {
      return revenueThisMonth(db, anchorDate);
    }
    return invalid("invalid_intent", `Unsupported seller-bi intent: ${commandCode}`);
  } finally {
    db.close();
  }
}

File: Archive/python/scripts/lib/seller_bi.py
"""Seller BI runtime for owner-facing metrics over orders and payments.

This module implements the first executable Seller BI slice for Purr Suite v1.
It exposes two owner-only read commands, both anchored by an explicit host
provided date so demo playback and integration tests stay deterministic.
"""

from __future__ import annotations

import sqlite3
from contextlib import closing
from datetime import date, timedelta
from typing import Any

from scripts.db.sqlite import DEFAULT_DB_PATH, ensure_database
from scripts.lib.money import format_minor_amount


COMMAND_ALIASES = {
    "seller-bi.sales_today": "seller_bi.sales_today",
    "seller-bi.revenue_this_month": "seller_bi.revenue_this_month",
}
SELLER_BI_COMMANDS = {
    "seller_bi.sales_today",
    "seller_bi.revenue_this_month",
}
SALES_COUNTABLE_ORDER_STATUSES = ("paid", "refunded", "fulfilled")


def _invalid(status: str, reply: str) -> dict[str, Any]:
    """Return a stable seller-bi runtime error payload."""

    return {"status": status, "reply": reply}


def _normalize_text(value: object, field_name: str) -> str:
    """Normalize one required text field and reject blank values."""

    normalized = str(value or "").strip()
    if not normalized:
        raise ValueError(f"{field_name} is required")
    return normalized


def _require_params_mapping(params: object) -> dict[str, Any]:
    """Validate the shared `params` envelope for seller-bi runtime."""

    if params is None:
        return {}
    if not isinstance(params, dict):
        raise ValueError("params must be an object")
    return dict(params)


def _canonical_command_code(value: object) -> str:
    """Normalize runtime command codes and fold supported legacy aliases."""

    normalized = str(value or "").strip().lower()
    return COMMAND_ALIASES.get(normalized, normalized)


def _parse_anchor_date(params: dict[str, Any]) -> date:
    """Parse the host-provided anchor date used for relative metric windows."""

    raw = _normalize_text(params.get("anchor_date"), "params.anchor_date")
    try:
        return date.fromisoformat(raw)
    except ValueError as exc:
        raise ValueError("params.anchor_date must be an ISO date like YYYY-MM-DD") from exc


def _day_window(anchor_date: date) -> tuple[str, str]:
    """Return the day window as SQLite-compatible timestamp bounds."""

    start = anchor_date.isoformat() + " 00:00:00"
    end = (anchor_date + timedelta(days=1)).isoformat() + " 00:00:00"
    return start, end


def _month_window(anchor_date: date) -> tuple[str, str]:
    """Return the month window as SQLite-compatible timestamp bounds."""

    month_start = anchor_date.replace(day=1)
    if month_start.month == 12:
        next_month_start = month_start.replace(year=month_start.year + 1, month=1)
    else:
        next_month_start = month_start.replace(month=month_start.month + 1)
    return (
        month_start.isoformat() + " 00:00:00",
        next_month_start.isoformat() + " 00:00:00",
    )


def _fetch_active_owner(
    conn: sqlite3.Connection, *, channel: str, external_user_id: str
) -> sqlite3.Row | None:
    """Return the active owner row for the caller, if one exists."""

    return conn.execute(
        """
        SELECT *
        FROM identities
        WHERE channel = ? AND external_user_id = ? AND is_active = 1 AND role = 'owner'
        LIMIT 1
        """,
        (channel, external_user_id),
    ).fetchone()


def _sales_today(conn: sqlite3.Connection, *, anchor_date: date) -> dict[str, Any]:
    """Count orders that first reached paid state during the anchored day."""

    window_start, window_end_exclusive = _day_window(anchor_date)
    row = conn.execute(
        """
        SELECT COUNT(*) AS sales_count
        FROM orders
        WHERE status IN (?, ?, ?)
          AND paid_at IS NOT NULL
          AND paid_at >= ?
          AND paid_at < ?
        """,
        (
            SALES_COUNTABLE_ORDER_STATUSES[0],
            SALES_COUNTABLE_ORDER_STATUSES[1],
            SALES_COUNTABLE_ORDER_STATUSES[2],
            window_start,
            window_end_exclusive,
        ),
    ).fetchone()
    sales_count = int(row["sales_count"]) if row is not None else 0
    return {
        "status": "computed",
        "reply": f"Sales count for {anchor_date.isoformat()} is {sales_count}.",
        "metric_code": "seller_bi.sales_today",
        "anchor_date": anchor_date.isoformat(),
        "window_start": window_start,
        "window_end_exclusive": window_end_exclusive,
        "sales_count": sales_count,
    }


def _revenue_this_month(conn: sqlite3.Connection, *, anchor_date: date) -> dict[str, Any]:
    """Return net paid revenue for the anchored month, grouped by currency."""

    window_start, window_end_exclusive = _month_window(anchor_date)
    month = anchor_date.strftime("%Y-%m")
    rows = conn.execute(
        """
        SELECT currency, SUM(amount_minor) AS amount_minor
        FROM payments
        WHERE status = 'paid'
          AND paid_at IS NOT NULL
          AND paid_at >= ?
          AND paid_at < ?
        GROUP BY currency
        ORDER BY currency ASC
        """,
        (window_start, window_end_exclusive),
    ).fetchall()
    revenue_rows = [
        {
            "currency": str(row["currency"]),
            "amount_minor": int(row["amount_minor"]),
            "display_amount": format_minor_amount(
                int(row["amount_minor"]),
                str(row["currency"]),
            ),
        }
        for row in rows
    ]
    currency_count = len(revenue_rows)
    if currency_count == 0:
        reply = f"No net paid revenue found for {month}."
    elif currency_count == 1:
        reply = f"Revenue for {month} is {revenue_rows[0]['display_amount']}."
    else:
        reply = f"Revenue for {month} spans {currency_count} currencies."
    return {
        "status": "computed",
        "reply": reply,
        "metric_code": "seller_bi.revenue_this_month",
        "anchor_date": anchor_date.isoformat(),
        "month": month,
        "window_start": window_start,
        "window_end_exclusive": window_end_exclusive,
        "currency_count": currency_count,
        "multi_currency": currency_count > 1,
        "revenue_rows": revenue_rows,
    }


def handle_seller_bi(context: dict[str, Any]) -> dict[str, Any]:
    """Handle normalized seller-bi runtime commands."""

    command_code = _canonical_command_code(context.get("command_code"))
    if command_code not in SELLER_BI_COMMANDS:
        return _invalid("invalid_intent", f"Unsupported seller-bi intent: {command_code}")

    user = context.get("user") or {}
    params = context.get("params")
    channel = str(context.get("channel") or "telegram").strip() or "telegram"
    external_user_id = str(user.get("external_user_id") or "").strip()
    if not external_user_id:
        return _invalid("invalid_input", "user.external_user_id is required")

    try:
        normalized_params = _require_params_mapping(params)
        anchor_date = _parse_anchor_date(normalized_params)
    except ValueError as exc:
        return _invalid("invalid_input", str(exc))

    runtime = context.get("runtime") or {}
    db_path = str(runtime.get("db_path") or DEFAULT_DB_PATH)
    with closing(ensure_database(db_path)) as conn:
        owner = _fetch_active_owner(
            conn, channel=channel, external_user_id=external_user_id
        )
        if owner is None:
            return _invalid(
                "forbidden",
                "Only the active owner can query seller-bi metrics.",
            )

        if command_code == "seller_bi.sales_today":
            return _sales_today(conn, anchor_date=anchor_date)
        if command_code == "seller_bi.revenue_this_month":
            return _revenue_this_month(conn, anchor_date=anchor_date)

    return _invalid("invalid_intent", f"Unsupported seller-bi intent: {command_code}")
